swift3/s3_token_middleware.py

```diff
@@ -197,21 +199,34 @@ def _deny_request(self, code):
 
     def _json_request(self, creds_json):
         headers = {'Content-Type': 'application/json'}
-        try:
-            response = self.session.post(
-                '%s/v2.0/s3tokens' % self._request_uri, headers=headers,
-                data=creds_json, verify=self._verify, timeout=self._timeout)
-        except requests.exceptions.Timeout as e:
-            self._logger.info('HTTP timeout: %s', e)
-            raise self._deny_request('ServiceUnavailable')
-        except requests.exceptions.RequestException as e:
-            self._logger.info('HTTP connection exception: %s', e)
-            raise self._deny_request('InvalidURI')
-
-        if response.status_code < 200 or response.status_code >= 300:
-            self._logger.debug('Keystone reply error: status=%s reason=%s',
-                               response.status_code, response.reason)
-            raise self._deny_request('AccessDenied')
+        for attempt in range(self._max_attempts):
+            try:
+                response = self.session.post(
+                    '%s/v2.0/s3tokens' % self._request_uri, headers=headers,
+                    data=creds_json, verify=self._verify,
+                    timeout=self._timeout)
+            except requests.exceptions.Timeout as e:
+                self._logger.info('HTTP timeout: %s', e)
+                raise self._deny_request('ServiceUnavailable')
+            except httplib.BadStatusLine as e:
+                # See https://github.com/requests/requests/issues/2364
+                if attempt + 1 >= self._max_attempts:
+                    raise
+                self._logger.warning('HTTP request BadStatusLine, retrying')
+                continue
+            except requests.exceptions.RequestException as e:
+                self._logger.info('HTTP connection exception: %s', e)
+                raise self._deny_request('InvalidURI')
+
+            if response.status_code >= 500:
+                self._logger.warning('HTTP request returned %s, retrying',
+                                     response.status_code)
+                continue
```

Comment:
If the last attempt has a status code greater than 500, no exception is raised.

Reply:
I thought the response status would be sufficient. But you are right, we must raise an exception.